fix(rpc): Refactor `getrawtransaction` & RPC error handling

upbqdn opened the pull request:
## Motivation

Close #8945.

### Specifications & References

- https://zcash.github.io/rpc/getrawtransaction.html
- https://github.com/zcash/zcash/blob/99ad6fdc3a549ab510422820eea5e5ce9f60a5fd/src/rpc/rawtransaction.cpp#L340

## Solution

- Fix the bugs described in #8744. Note that `zcashd` stores transactions in orphaned blocks and responds to queries requesting such transactions with `height` to -1. Since Zebra doesn't keep orphaned blocks below the rollback limit, it now returns the error _"No such mempool or blockchain transaction"_. If we wanted to match `zcashd`, we'd need to start storing orphaned blocks.
- Fix the error codes for `getrawtransaction` and some other RPCs.
- Related cleanups:
  - Refactor the error handling for RPCs.
  - Add references to the original error codes.
  - Simplify prop tests and run them with arbitrary `Network`s.
  - [Don't derive Serialize & Deserialize for txids](https://github.com/ZcashFoundation/zebra/pull/9049/commits/c9088f8f97ea70659ae6f333f2fcb7735ac451b6)

### Tests

Update and add

- snapshot tests,
- unit tests, and
- prop tests

to check the new error codes.

### Follow-up Work

- None

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

upbqdn committed:
clean-up: simplify the def of `MapServerError`

upbqdn committed:
Use `HexData` instead of `String` for TXIDs

upbqdn committed:
Remove a redundant test

We don't need such a test anymore because the deserialization is handled
by Serde now.

upbqdn committed:
Adjust tests for using `HexData`

upbqdn committed:
Make `height` and `confirmations` optional

upbqdn committed:
Use legacy error codes

upbqdn committed:
fmt

upbqdn committed:
Remove unneeded error codes

upbqdn committed:
Remove `zebra-rpc/src/constants.rs`

upbqdn committed:
Rename `MapServerError` to `MapError`

upbqdn committed:
Rename `OkOrServerError` to `OkOrError`

upbqdn committed:
Allow specifying error codes when mapping errors

upbqdn committed:
Allow setting error codes when mapping options

upbqdn committed:
Use the right error code for `getrawtransaction`

upbqdn committed:
fmt

upbqdn committed:
Add docs for the error conversion traits

upbqdn committed:
Refactor the error handling for `getblock`

upbqdn committed:
Refactor error handling in `sendrawtransaction`

upbqdn committed:
Refactor the error handling for `getblock`

upbqdn committed:
Update the error handling for `getrawtransaction`

upbqdn committed:
Refactor error handling for `z_gettreestate`

upbqdn committed:
Refactor the error handling for address parsing

upbqdn committed:
Refactor the error handling for getrawtransaction

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Update `z_gettreestate` snapshots

upbqdn committed:
Cosmetics

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Refactor error handling in `getblock`

upbqdn committed:
Refactor error handling in `getblockheader`

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Simplify `getrawtransaction`

upbqdn committed:
Check errors for `getrawtransaction`

upbqdn committed:
fmt

upbqdn committed:
Simplify proptests

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Fix unit tests for `getaddresstxids`

upbqdn committed:
Fix unit tests for `getaddressutxos`

upbqdn committed:
fix docs

upbqdn committed:
Update snapshots for `getrawtransaction`

arya2 commented on zebra-rpc/src/server/error.rs:
```diff
@@ -0,0 +1,105 @@
+        code: impl Into<jsonrpc_core::ErrorCode>,
+        message: impl ToString,
+    ) -> std::result::Result<T, jsonrpc_core::Error>;
+}
```
Error constructor methods for commonly-used error codes (or just `Misc`) could be nice here.

```suggestion
/// A trait for mapping errors to [`jsonrpc_core::Error`].
pub(crate) trait MapError<T>: Sized {
    /// Maps errors to [`jsonrpc_core::Error`] with a specific error code.
    fn map_error(
        self,
        code: impl Into<jsonrpc_core::ErrorCode>,
    ) -> std::result::Result<T, jsonrpc_core::Error>;

    /// Maps errors to [`jsonrpc_core::Error`] with a [`LegacyCode::Misc`] error code.
    fn map_misc_error(self) -> std::result::Result<T, jsonrpc_core::Error> {
        self.map_error(LegacyCode::Misc)
    }
}

/// A trait for conditionally converting a value into a `Result<T, jsonrpc_core::Error>`.
pub(crate) trait OkOrError<T>: Sized {
    /// Converts the implementing type to `Result<T, jsonrpc_core::Error>`, using an error code and
    /// message if conversion is to `Err`.
    fn ok_or_error(
        self,
        code: impl Into<jsonrpc_core::ErrorCode>,
        message: impl ToString,
    ) -> std::result::Result<T, jsonrpc_core::Error>;

    /// Converts the implementing type to `Result<T, jsonrpc_core::Error>`, using a [`LegacyCode::Misc`] error code.
    fn ok_or_misc_error(
        self,
        message: impl ToString,
    ) -> std::result::Result<T, jsonrpc_core::Error> {
        self.ok_or_error(LegacyCode::Misc, message)
    }
}
```

arya2 commented on zebra-rpc/src/methods.rs:
```diff
@@ -1116,73 +1149,80 @@ where
     fn get_raw_transaction(
         &self,
-        txid_hex: String,
+        HexData(txid): HexData,
```
This could be typed as a `transaction::Hash` (which implements `ToHex` & `FromHex`), though it may return the wrong error code that way.

upbqdn replied:
Thanks, I did that in [bdb8372](https://github.com/ZcashFoundation/zebra/pull/9049/commits/bdb837284d7fe8089b25e19d180b68d82a2b3ef6). The error code is not compatible with `zcashd`, but it wasn't anyway if the submitted string wasn't hex. I'll try to figure out if we need compatibility here and open another PR if we do.

upbqdn replied:
I returned to using `String`s for the TXIDs even though they provide a weaker semantic abstraction. The reason is that `String`s are exactly what the RPC gets from the client. This means we can have greater test coverage (and error control). Using `transaction::Hash` felt great, but then I had to submit that directly to the RPC instead of a `String`, and we couldn't test deserialization properly, which led to bugs like https://github.com/ZcashFoundation/zebra/pull/9049#issuecomment-2536303946.

arya2 commented on zebra-rpc/src/methods.rs:
```diff
@@ -1116,73 +1149,80 @@ where
+            match state
                 .ready()
-                .and_then(|service| service.call(request))
+                .and_then(|service| service.call(zebra_state::ReadRequest::Transaction(txid)))
```
Optional: To match zcashd's output when a transaction is on a side chain, we either need a new request to check for transactions on side chains, or to update this request so that it checks side chains for the transaction id too and returns an enum variant indicating whether the transaction was found on the best chain or a side chain.

Update: Let's skip this or do it in a follow-up PR if work hasn't already started to try and get https://github.com/ZcashFoundation/zebra/pull/9059 merged sooner.

arya2 commented on zebra-rpc/src/methods.rs:
```diff
@@ -733,200 +746,212 @@ where
-        let verbosity = verbosity.unwrap_or(DEFAULT_GETBLOCK_VERBOSITY);
 
         async move {
-            let hash_or_height: HashOrHeight = hash_or_height.parse().map_server_error()?;
```
These changes to the `get_block()` method may conflict with those in https://github.com/ZcashFoundation/zebra/pull/9006.

upbqdn replied:
Let's resolve those when we merge either of the PRs.

arya2 reviewed (commented):
This is looking good so far, I left an optional suggestion about matching zcashd's output when a transaction is found in a side chain. 

I'd like to review the changes to the `get_block()` method against #9006 if we can target this PR against the `get-block` branch (or wait for #9006 to merge and rebase off of main).

upbqdn committed:
Update zebra-rpc/src/server/error.rs

Co-authored-by: Arya <aryasolhi@gmail.com>

oxarbitrage commented on zebra-rpc/src/server/error.rs:
```diff
@@ -0,0 +1,118 @@
+        code: impl Into<jsonrpc_core::ErrorCode>,
+    ) -> std::result::Result<T, jsonrpc_core::Error>;
+
+    /// Maps errors to [`jsonrpc_core::Error`] with a [`LegacyCode::Misc`] error code.
+    fn map_misc_error(self) -> std::result::Result<T, jsonrpc_core::Error> {
+        self.map_error(LegacyCode::Misc)
+    }
```
Unused method. I got a warning building locally.
```suggestion
```

upbqdn replied:
That was just an interim inconsistency. Addressed in [ed69359](https://github.com/ZcashFoundation/zebra/pull/9049/commits/ed693598c69a0ef3b1e92e982783e0992d61399b).

oxarbitrage commented on zebra-rpc/src/server/error.rs:
```diff
@@ -0,0 +1,118 @@
+        message: impl ToString,
+    ) -> std::result::Result<T, jsonrpc_core::Error>;
+
+    /// Converts the implementing type to `Result<T, jsonrpc_core::Error>`, using a [`LegacyCode::Misc`] error code.
+    fn ok_or_misc_error(
+        self,
+        message: impl ToString,
+    ) -> std::result::Result<T, jsonrpc_core::Error> {
+        self.ok_or_error(LegacyCode::Misc, message)
+    }
```
```suggestion
```

upbqdn replied:
That was just an interim inconsistency. Addressed in [ed69359](https://github.com/ZcashFoundation/zebra/pull/9049/commits/ed693598c69a0ef3b1e92e982783e0992d61399b).

oxarbitrage commented on zebra-rpc/src/methods.rs:
```diff
@@ -733,200 +746,212 @@ where
+                    // # Performance
+                    //
+                    // This RPC is used in `lightwalletd`'s initial sync of 2 million blocks,
+                    // so it needs to load block data very efficiently.
```
```suggestion
```

This comment is repeated. I think this one can be removed but please double check before.

upbqdn replied:
I didn't find similar comments. Could you share which ones you had in mind?

arya2 replied:
I think this comment is true for verbosity 1, but not for verbosity 0 here as lightwalletd doesn't use getblock with verbosity 0.

There is a note about performance [here](https://github.com/ZcashFoundation/zebra/pull/9049/files/475dbc9df3f041d5b6722539d136f8d2b3e56dfe#diff-09daf8f97cb59393d0c24039a43290af0772271b820987ef1a540404a78c44dbR786-R793) for the verbosity 1 arm of this match statement.

upbqdn replied:
I can't apply this suggestion anymore, so I moved the note at the top of the fn in [2c342f7](https://github.com/ZcashFoundation/zebra/pull/9049/commits/2c342f7e1ad48e261ae6b32ebd423e83b34fe800). I moved it because the verbosity 1 arm is not in the code anymore.

arya2 replied:
> I can't apply this suggestion anymore

It's essentially the intended change, I'd describe that as having applied the suggestion, though I'm not sure if that's accurate.

upbqdn replied:
I just meant that the "apply" button here on GitHub wasn't active anymore, so I did the suggested changes locally. The downside is that @oxarbitrage didn't get any credit in the commit history.

oxarbitrage commented on zebra-rpc/src/server/error.rs:
```diff
@@ -0,0 +1,118 @@
+
+/// Bitcoin RPC error codes
+///
+/// Drawn from <https://github.com/zcash/zcash/blob/master/src/rpc/protocol.h#L32-L80>.
```
```suggestion
/// Drawn from <https://github.com/zcash/zcash/blob/99ad6fdc3a549ab510422820eea5e5ce9f60a5fd/src/rpc/protocol.h#L32-L80>.
```

It is maybe safer to link to a commit.

oxarbitrage commented on zebra-rpc/src/methods/tests/snapshots/getrawtransaction_invalid_txid@mainnet_10.snap:
```diff
@@ -0,0 +1,11 @@
+---
+source: zebra-rpc/src/methods/tests/snapshot.rs
+expression: rsp
+snapshot_kind: text
```
Unsure if we should/can just remove this line from the snapshots.

upbqdn replied:
You mean the `snapshot_kind`? What would be the motivation?

oxarbitrage replied:
Only the snapshots of this PR has the `snapshot_kind`.

upbqdn replied:
I think it's fine as it is.

oxarbitrage commented on zebra-rpc/src/server/error.rs:
```diff
@@ -0,0 +1,118 @@
+    #[default]
+    Misc = -1,
+    /// Server is in safe mode, and command is not allowed in safe mode
+    ForbiddenBySafeMode = -2,
```
We could just remove the ones we are not using and suggest to be added on demand. However, as i am not sure what is better, i am ok keeping it like it is.

upbqdn replied:
Not a biggie, so I'll leave it as it is.

oxarbitrage reviewed (commented):
this looks good to me, i left a few minor comments and suggestions, also we need to fix conflicts.

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Use `transaction::Hash` instead of `HexData`

upbqdn committed:
Simplify error handling

upbqdn committed:
Update zebra-rpc/src/server/error.rs

Co-authored-by: Alfredo Garcia <oxarbitrage@gmail.com>

arya2 reviewed (dismissed):
Looks good to me, it would be nice to apply [this suggestion](https://github.com/ZcashFoundation/zebra/pull/9049#discussion_r1876731824) but it could also be removed in the next PR so I don't think it's a blocker.

upbqdn committed:
Move a note on performance

upbqdn committed:
Fix a typo

arya2 commented:
The lightwalletd tests seem to be required now, they should pass once https://github.com/ZcashFoundation/zebra/pull/9052 has merged. There was a timeout here: https://github.com/ZcashFoundation/zebra/actions/runs/12265821808/job/34222718084?pr=9049#step:12:2016, 

but the same tests passed with rust stable so it was likely spurious.

arya2 commented:
@Mergify update

mergify[bot] committed:
Merge branch 'main' into fix-getrawtx

conradoplg reviewed (changes requested):
I think there might be a bug, I'm looking into it

conradoplg commented:
Yep, e.g. `zcash-cli -conf=$PWD/zcash.conf getrawtransaction 851bf6fbf7a976327817c738c489d7fa657752445430922d94c983c0b9ed4609 1` is returning

```
error code: -1
error message:
Invalid params: invalid type: string "1e4842fd4962d395cd9905b7030fd5cb54c8c30148468c600416e4f14a264d2c", expected an array of length 32.
```

on `main` it returns

```
{
  "hex": "01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff025100ffffffff0250c30000000000002321027a46eb513588b01b37ea24303f4b628afd12cc20df789fede0921e43cad3e875acd43000000000000017a9147d46a730d31f97b1930d3368a967c309bd4d136a8700000000",
  "height": 1,
  "confirmations": 2740688
}
```

upbqdn commented:
Thanks for spotting the bug. I'll switch back to `String`s and deserialize them manually instead of using `transaction::Hash` directly.

upbqdn committed:
Use `String` instead of `transaction::Hash`

upbqdn committed:
Adjust and add proptests

upbqdn committed:
Merge branch 'main' into fix-getrawtx

upbqdn committed:
Reintroduce snapshots for invalid TXIDs

upbqdn committed:
Merge branch 'fix-getrawtx' of github.com:ZcashFoundation/zebra into fix-getrawtx

upbqdn commented:
> I think there might be a bug, I'm looking into it

Fixed.

upbqdn committed:
Don't derive `Serialize` & `Deserialize` for txids

Deriving `serde::Serialize` & `serde::Deserialize` for
`transaction::Hash` was superfluous, and we didn't need it anywhere in
the code.
